include/picongpu/algorithms/FieldToParticleInterpolation.hpp

```diff
@@ -63,6 +64,22 @@ namespace picongpu
         static constexpr int begin = -supp / 2 + (supp + 1) % 2;
         static constexpr int end = begin + supp - 1;
 
+        HDINLINE auto getShapeArray(float3_X const& pos) const
+        {
+            pmacc::memory::Array<shapes::Cached<AssignmentFunction>, 3> result;
+            result[0] = shapes::Cached<AssignmentFunction>(pos.x(), true);
+            result[1] = shapes::Cached<AssignmentFunction>(pos.y(), true);
+            result[2] = shapes::Cached<AssignmentFunction>(pos.z(), true);
+            return result;
+        };
+
+        HDINLINE auto getShapeArray(float2_X const& pos) const
+        {
+            pmacc::memory::Array<shapes::Cached<AssignmentFunction>, 2> result;
+            result[0] = shapes::Cached<AssignmentFunction>(pos.x(), true);
+            result[1] = shapes::Cached<AssignmentFunction>(pos.y(), true);
+            return result;
+        };
```

Comment:
I think at least this part is easy to express generically for 2d and 3d and not have two copy-pasted functions?

Reply:
IN general, it is depending on the field.